File: insert_postgres.py
import psycopg2
import time

# Konfigurasi koneksi PostgreSQL
DB_NAME = "coppegadaian"
DB_USER = "postgres"
DB_PASSWORD = "admin"
DB_HOST = "localhost"
CSV_FILE = "dummy_data.csv"  # Sesuaikan dengan path CSV

def connect_db():
    """Membuka koneksi ke PostgreSQL."""
    try:
        conn = psycopg2.connect(
            dbname=DB_NAME, user=DB_USER, password=DB_PASSWORD, host=DB_HOST
        )
        print("Connected to PostgreSQL!")
        return conn
    except Exception as e:
        print("Error connecting to database:", e)
        exit(1)

def create_table(cursor):
    """Membuat tabel jika belum ada."""
    cursor.execute(
        """
        CREATE TABLE IF NOT EXISTS userspy (
            id SERIAL PRIMARY KEY,
            name VARCHAR(50),
            email VARCHAR(100),
            created_at TIMESTAMP DEFAULT NOW(),
            price FLOAT,
            status VARCHAR(20),
            description TEXT
        )
        """
    )

def insert_data():
    """Mengimpor data dari CSV ke PostgreSQL menggunakan COPY."""
    conn = connect_db()
    cursor = conn.cursor()

    # Buat tabel jika belum ada
    create_table(cursor)
    conn.commit()
    print("Table users checked/created.")

    # Bersihkan tabel (Opsional: Hapus jika tidak ingin truncate data lama)
    cursor.execute("TRUNCATE TABLE userspy RESTART IDENTITY CASCADE")
    conn.commit()
    print("Table users truncated.")

    # Proses Insert
    print("Starting INSERT process using COPY...")
    start_time = time.time()

    try:
        with open(CSV_FILE, "r", encoding="utf-8") as f:
            next(f)  # Skip header
            cursor.copy_expert(
                "COPY userspy(id,name, email, created_at, price, status, description) FROM STDIN WITH (FORMAT csv, HEADER true)", f
            )
        conn.commit()
        print("Insert completed using COPY!")

    except Exception as e:
        print("Error:", e)
        conn.rollback()

    end_time = time.time()
    print(f"Total time taken: {end_time - start_time:.2f} seconds")

    cursor.close()
    conn.close()

if __name__ == "__main__":
    insert_data()

File: insert_postgres.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <libpq-fe.h>

void check_exec_status(PGconn *conn, PGresult *res, const char *message) {
    if (PQresultStatus(res) != PGRES_COMMAND_OK) {
        fprintf(stderr, "ERROR: %s: %s\n", message, PQerrorMessage(conn));
        PQclear(res);
        PQfinish(conn);
        exit(1);
    }
}

int main() {
    // Catat waktu mulai
    clock_t start_time = clock();

    PGconn *conn = PQconnectdb("dbname=coppegadaian user=postgres password=admin host=localhost");
    
    if (PQstatus(conn) != CONNECTION_OK) {
        fprintf(stderr, "Connection failed: %s\n", PQerrorMessage(conn));
        PQfinish(conn);
        return 1;
    }
    printf("Connected to PostgreSQL!\n");

    // Pastikan tabel sudah ada
    PGresult *res = PQexec(conn, 
        "CREATE TABLE IF NOT EXISTS users ("
        "id SERIAL PRIMARY KEY, "
        "name VARCHAR(50), "
        "email VARCHAR(100), "
        "created_at TIMESTAMP, "
        "price FLOAT, "
        "status VARCHAR(20), "
        "description TEXT)"
    );
    check_exec_status(conn, res, "Table creation failed");
    PQclear(res);

    printf("Starting INSERT process using COPY...\n");

    // Load CSV dari file Python
    char copy_sql[512];
    snprintf(copy_sql, sizeof(copy_sql), 
        "COPY users (id,name, email, created_at, price, status, description) "
        "FROM STDIN WITH (FORMAT csv, HEADER true)");

    res = PQexec(conn, copy_sql);
    if (PQresultStatus(res) != PGRES_COPY_IN) {
        fprintf(stderr, "COPY command failed: %s\n", PQerrorMessage(conn));
        PQclear(res);
        PQfinish(conn);
        return 1;
    }
    PQclear(res);

    // Kirim data dari file CSV ke PostgreSQL
    FILE *csv = fopen("clean_data.csv", "r");
    if (!csv) {
        fprintf(stderr, "Error opening CSV file.\n");
        PQfinish(conn);
        return 1;
    }

    char buffer[8192];
    while (fgets(buffer, sizeof(buffer), csv)) {
        if (PQputCopyData(conn, buffer, strlen(buffer)) != 1) {
            fprintf(stderr, "Error sending data to PostgreSQL: %s\n", PQerrorMessage(conn));
            fclose(csv);
            PQputCopyEnd(conn, "Error during COPY");
            PQfinish(conn);
            return 1;
        }
    }
    fclose(csv);
    
    // Selesaikan COPY
    if (PQputCopyEnd(conn, NULL) != 1) {
        fprintf(stderr, "Error finishing COPY: %s\n", PQerrorMessage(conn));
        PQfinish(conn);
        return 1;
    }

    // Catat waktu selesai
    clock_t end_time = clock();
    double elapsed_time = (double)(end_time - start_time) / CLOCKS_PER_SEC;

    printf("Insert completed using COPY!\n");
    printf("Total time taken: %.2f seconds\n", elapsed_time);

    PQfinish(conn);
    return 0;
}
